language/move-vm/runtime/src/data_cache.rs

```diff
@@ -158,7 +158,13 @@ impl<'r, 'l, R: RemoteCache> TransactionDataCache<'r, 'l, R> {
     }
 
     pub(crate) fn num_mutated_accounts(&self) -> u64 {
-        self.account_map.keys().len() as u64
+        let mut total_mutate_accounts: u64 = 0;
+        for (_, entry) in self.account_map.iter() {
+            if entry.data_map.values().any(|(_, v)| v.is_mutated()) {
+                total_mutate_accounts += 1;
+            }
+        }
+        total_mutate_accounts
     }
```

Comment:
To me, `num_mutated_accounts` has always been a relatively weird API to expose from the VM/Session, but I understand there's some tricky logic we need to deal with at the call site...

@runtian-zhou could you remind me why we don't calculate this from the transaction effects? Is it because we need to know the number before going into the epilogue, where gas distribution is handled?